lookout/style/typos/utils.py

```diff
@@ -0,0 +1,177 @@
+
+
+def add_context_info(data: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    Split context of identifier on before and after part
+    :param data: DataFrame. Column "token_split" will be used for
+        creating context info if present
+    :return: Provided data with added columns "before" and "after", containing lists
+        of corresponding contexts tokens
+    """
+    if "before" in data.columns and "after" in data.columns:
+        return data
+
+    tokens = list(data.typo)
+
+    if "token_split" in data.columns:
+        token_split = list(data.token_split)
+        before = []
+        after = []
+
+        for row_number in range(len(data)):
+            if "token_split" in data.columns:
+                split = token_split[row_number]
+                if isinstance(split, str):
+                    split = split.split()
+                index = split.index(tokens[row_number])
+                before.append(split[:index])
+                after.append(split[index + 1:])
+
+    else:
+        before = [[] for _ in range(len(data))]
+        after = [[] for _ in range(len(data))]
+
+    data["before"] = before
+    data["after"] = after
+
+    return data.infer_objects()
+
+
+def rank_candidates(candidates: pandas.DataFrame, pred_proba: list, n_candidates: int = None,
+                    return_all: bool = True):
+    """
+    Rank candidates for typos correction based on correctness probabilities.
+    :param candidates: Dataframe with columns "id", "typo", "candidate"
+        and indexed by range(len(pred_proba))
+    :param pred_proba: Array of probabilities of correctness of every candidate
+    :param n_candidates: Number of most probably correct candidates to return for each typo
+    :param return_all: False to return corrections only for
+        corrected in the first candidate typos
+    :return: Dictionary {id : [[candidate, correct_prob]]}, candidates are sorted
+        by correct_prob in a descending order.
+    """
+    suggestions = {}
+    corrections = []
+    for i in range(len(pred_proba)):
+        index = candidates.loc[i, "id"]
+        corrections.append([candidates.loc[i, "candidate"], pred_proba[i]])
+
+        if i < len(pred_proba) - 1 and candidates.loc[i + 1, "id"] == index:
+            continue
+
+        corrections = list(sorted(corrections, key=lambda x: -x[1]))
+        typo = candidates.loc[i, "typo"]
+        if corrections[0][0] != typo:
+            suggestions[index] = (corrections if n_candidates is None
+                                  else corrections[:n_candidates])
+        elif return_all:
+            suggestions[index] = [[typo, 1.0]]
+        corrections = []
+
+    return suggestions
+
+
+def filter_suggestions(typos: pandas.DataFrame, suggestions: dict, n_candidates: int=1,
+                       return_all: bool=False) -> dict:
+    """
+    Filter corrections suggestions
+    :param typos: DataFrame, containing column "typo"
```

Comment:
Style - comments should end with a dot - https://github.com/src-d/style-analyzer/blob/master/lookout/style/format/features.py#L223